html/semantics/embedded-content/the-video-element/resize-during-playback.html

```diff
@@ -0,0 +1,46 @@
+<!doctype html>
+<html>
+<head>
+<title>video element resizing during playback</title>
+<link rel="help" href="https://html.spec.whatwg.org/multipage/media.html#concept-video-intrinsic-width">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+</head>
+<body>
+<div id="log"></div>
+<script>
+setup(() => {
+  let anyFormatSupported = false;
+  for (const format of ['mp4', 'webm']) {
+    const video = document.createElement('video');
+    // TODO: skip subtests after https://github.com/web-platform-tests/rfcs/pull/16
+    if (video.canPlayType(`video/${format}`)) {
+      anyFormatSupported = true;
+      promise_test(async (t) => {
+        // Load the video and wait for initial resize event.
+        video.src = `/media/400x300-red-resize-300x150-green.${format}`;
+        video.preload = 'auto';
+        document.body.appendChild(video);
+        const eventWatcher = new EventWatcher(t, video, ['resize', 'playing', 'error', 'ended']);
+        await eventWatcher.wait_for(['resize']);
+        assert_equals(video.videoWidth, 400, 'width after first resize event');
```

Comment:
Does a specification state that the _first_ `resize` event MUST result in the `width` being set to a specific value relative to the underlying video track?

Reply:
An possible adjustment to the code would be to 
1. Check if `videoWidth` is not equal to expected at first `resize` event
2. If `false` await at least one more `resize` event. If a specification defines a time `resize` event should occur, fail test after N time, else check if 1. is true at subsequent `resize` event(s), when true pass test.

Reply:
It does

> For video elements, set the videoWidth and videoHeight attributes, and queue a task to fire an event named resize at the media element.

https://html.spec.whatwg.org/multipage/media.html#loading-the-media-resource:event-media-resize

Reply:
That language does not specify that the `videoWidth` and `videoHeight` attributes MUST be set to the exact pixel dimensions of the underlying frame at the _first_ resize event.

If the language in the standard needs to be changed to state that then a PR should be filed.

Reply:
@zcorpan During testing with recording variable pixel dimension input videos to a single video found that it was necessary to check if the pixel dimensions had been set to the expected value at that specific `resize` event, a `resize` event can occur and the current `<video>` displayed pixel dimensions can be captured _between_

> Note
>> Further resize events will be fired if the dimensions subsequently change.
 
the previous and expected values, which required testing to determine that fact and compose code to try handle those cases, see the linked plnkr, above, which intententionally does not attempt to fix the result completely.

Reply:
> That language does not specify that the `videoWidth` and `videoHeight` attributes MUST be set to the exact pixel dimensions of the underlying frame at the _first_ resize event.

Yes, it does. As far as I can tell, this is the place in the spec that would fire the first `resize` event. The algorithm is invoked with a "must" here:

> The networking task source tasks to process the data as it is being fetched must each immediately queue a task to run the first appropriate steps from the media data processing steps list below.

https://html.spec.whatwg.org/multipage/media.html#concept-media-load-resource

Since it says to set `videoWidth` and `videoHeight` and then queue a task to fire `resize`, they will be set before the `resize` event is actually fired.

Possibly the spec should say to update the [intrinsic width and intrinsic height](https://html.spec.whatwg.org/multipage/media.html#concept-video-intrinsic-width) based on the new information rather than setting readonly attributes, though.

Reply:
I guess it's possible that the intrinsic dimensions could change again before the first `resize` event is actually fired. Your test doesn't seem to change dimensions that rapidly, though.

Reply:
@zcorpan 

> Yes, it does. As far as I can tell, this is the place in the spec that would fire the first resize event. The algorithm is invoked with a "must" here:

>> The networking task source tasks to process the data as it is being fetched must each immediately queue a task to run the first appropriate steps from the media data processing steps list below.

Do not concur with that reading. 

The langauge is not specific as to _first_ and MUST and further that is not what actually occurs at the implementation, particularly when the input is, for example, a `CanvasCaptureMediaStreamTrack` where the frames can be drawn up to 60 times per second.

> I guess it's possible that the intrinsic dimensions could change again before the first `resize` event is actually fired. Your test doesn't seem to change dimensions that rapidly, though.

What definition of "rapidly" are you relying on? 

When the linked plnkr is run the _first_ `resize` event is not guranteed to be synchronous relevant to the `videoWidth` and `videoHeight` being any particular _value_. That is the observable result. And not the only means of producing such a result. The specification, as currently written, does not mandate that any particular `resize` event be directly related to the resulting value of _specific_ pixel dimensions. What _is_ stated is that subsequent `resize` events may occur.

Am not certain the specification _can_ be updated to state that the `videoWidth` and `videoHeight` must be equal to a _specific_ value. That does not currently happen.

How could one of the various decoders an implementation might be using for that codec possibly be guaranteed to already have _read_ the _next_ frame?

It would probably not be possible to implement a stricter standard. Only pointing out the current standard does not state any `resize` event MUST be consistent with the `target` `videoWidth` and `videoHeight` being _equal_ to any particular _value_. That fact cannot be changed or language in a comment inserted to state that mandate without editing the current specification to do so in fact. The standard does not state that right now.

Am only sharing what found during testing in the wild. Signed up to be a "member" of two of the bodies when it was suggested to do so to contribute to one of these specifications, fraudulent reasons for expulsion occurred, was banned, was not the first time an institution did that, for manufactured reasons that are trivial to prove are hypocritical, therefore am not necessarily impressed with the specification entity "member" process; take or leave what have found during field testing https://gist.githubusercontent.com/guest271314/1fcb5d90799c48fdf34c68dd7f74a304/raw/fa9edb61431b5bc7d7432335ca3ece5763f6209b/UseItBreakItFileBugsRequestFeaturesTestTestTestTestToThePointItBreaks.txt consider, and if helpful, apply to updating the relevant specifications and tests, or not.

Reply:
@zcorpan There are several things going on in the linked plnkr. Some background

1. Technically unrelated though in this case appears to have an impact on the effective result of `resize`, depending on which `resize` event is read. Chromium has a bug where if a `MediaStream` is not set as `srcObject` of an `HTMLMediaElement` `MediaRecorder` `dataavailable` event will output a `Blob` having `size` set to `0` [Issue 952700: MediaRecorder does not record MediaStream unless set as srcObject of <video> element](https://bugs.chromium.org/p/chromium/issues/detail?id=952700).
2. Without using `pause()` and `resume()` methods of `MediaRecorder` to await `resize` events and check `videoWidth` and `videoHeight` the resulting video (`Blob` => `Blob URL`) can output:  
`resulting video resize 384 288`
`resulting video resize 480 240`
`resulting video resize 640 360`
`resulting video resize 400 300`
`resulting video resize 1280 720`

    or
`resulting video resize 384 288`
`resulting video resize 576 432`
`resulting video resize 480 240`
`resulting video resize 640 360`
`resulting video resize 400 300`
`resulting video resize 1280 720`


    where the expected result is: 
`resize every resize 768 576`
`resize every resize 480 240`
`resize every resize 640 360`
`resize every resize 400 300`
`resize every resize 1280 720 // video has same pixel dimensions`



which means that the `<video>` is displaying and `MediaRecorder` is capturing `width` and `height` not equal to the underlying `videoWidth` and `videoHeight`.

Reading the current specification did re `resize` event, at least from persepecctive here, did not see where could assert the current language means that at the _**first**_ `resize` event the `videoWidth`, `videoHeight` (or `width` and `height`) of `<video>` **MUST** match the pixel dimensions of the underlying of the video track.

Filed more than one bug relevant to variable width and height decoding and recording at Chromium. 

If your reading of the specification is that the Chromium implementation is not specification compliant, kindly advise and will file a Chromium bug. 

--

Code modified from the linked plnkr to reproduce the above result

```
<!DOCTYPE html>
<html>
  <head>
    <title>
      Record MediaStream potentially including variable width and height
      MediaStreamTrack
    </title>
  </head>

  <body>
    <video id="video" autoplay controls></video>
    <script type="module">
      // MediaStream outputting #000000 video frames and audio silence
      import { mediaStream, audioContext, worker } from './export.js';
      let recorder, result;
      const videoStream = document.getElementById('video');
      console.log(mediaStream, audioContext, worker);
      (async () => {
        let handleChromiumVideoElementMediaRecorderBug;
        const [audioTrack, videoTrack] = ['audio', 'video'].map(kind =>
          mediaStream
            .getTracks()
            .find(({ kind: trackKind }) => trackKind === kind)
        );
        const fromLocalPeerConnection = new RTCPeerConnection();
        const toLocalPeerConnection = new RTCPeerConnection();
        const fromConnection = new Promise(resolve =>
          fromLocalPeerConnection.addEventListener(
            'icecandidate',
            async e => {
              console.log('from', e);
              try {
                resolve(
                  toLocalPeerConnection.addIceCandidate(
                    e.candidate ? e.candidate : null
                  )
                );
              } catch (e) {
                console.error(e);
              }
            },
            {
              once: true,
            }
          )
        );
        const toConnection = new Promise(resolve =>
          toLocalPeerConnection.addEventListener(
            'icecandidate',
            async e => {
              console.log('to', e);
              try {
                resolve(
                  fromLocalPeerConnection.addIceCandidate(
                    e.candidate ? e.candidate : null
                  )
                );
              } catch (e) {
                console.error(e);
              }
            },
            {
              once: true,
            }
          )
        );
        fromLocalPeerConnection.addEventListener('negotiationneeded', e => {
          console.log(e);
        });
        toLocalPeerConnection.addEventListener('negotiationneeded', e => {
          console.log(e);
        });
        let tracks = new Promise(resolve => {
          let ids = mediaStream.getTracks().map(({ id }) => id),
            i = 0;
          toLocalPeerConnection.addEventListener(
            'track',
            ({ track, streams: [stream] }) => {
              if (ids.includes(track.id) && ++i === ids.length) {
                resolve(stream);
              }
            }
          );
        });
        
        // Add initial audio and video MediaStreamTrack to PeerConnection, pass initial MediaStream
        const {
          sender: audioSender,
          receiver: audioReceiver,
        } = fromLocalPeerConnection.addTransceiver(audioTrack, {
          streams: [mediaStream],
        });
        const {
          sender: videoSender,
          receiver: videoReceiver,
        } = fromLocalPeerConnection.addTransceiver(videoTrack, {
          streams: [mediaStream],
        });
        console.log(audioSender, videoSender, audioReceiver, videoReceiver);
        const offer = await fromLocalPeerConnection.createOffer();
        await toLocalPeerConnection.setRemoteDescription(offer);
        await fromLocalPeerConnection.setLocalDescription(
          toLocalPeerConnection.remoteDescription
        );
        const answer = await toLocalPeerConnection.createAnswer();
        await fromLocalPeerConnection.setRemoteDescription(answer);
        await toLocalPeerConnection.setLocalDescription(
          fromLocalPeerConnection.remoteDescription
        );
        const stream = await tracks;
        // video.srcObject = stream;

        // When MediaStream is not set as srcObject Blob at MediaRecorder
        // dataavailable event size is 0
        const urls = await Promise.all(
          [
            {
              src:
                'https://upload.wikimedia.org/wikipedia/commons/a/a4/Xacti-AC8EX-Sample_video-001.ogv',
              from: 0,
              to: 4,
            },
            {
              src:
                'https://mirrors.creativecommons.org/movingimages/webm/ScienceCommonsJesseDylan_240p.webm#t=10,20',
              from: 10,
              to: 20,
            },
            {
              from: 55,
              to: 60,
              src:
                'https://nickdesaulniers.github.io/netfix/demo/frag_bunny.mp4',
            },
            {
              from: 0,
              to: 5,
              src:
                'https://raw.githubusercontent.com/w3c/web-platform-tests/master/media-source/mp4/test.mp4',
            },
            {
              from: 0,
              to: 5,
              src:
                'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerBlazes.mp4',
            },
            {
              from: 0,
              to: 5,
              src:
                'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerJoyrides.mp4',
            },
            {
              from: 0,
              to: 6,
              src:
                'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerMeltdowns.mp4#t=0,6',
            },
          ].map(async ({ from, to, src }) => {
            try {
              const request = await fetch(src);
              const blob = await request.blob();
              const blobURL = URL.createObjectURL(blob);
              const url = new URL(src);
              console.log(url.hash);
              return blobURL + (url.hash || `#t=${from},${to}`);
            } catch (e) {
              throw e;
            }
          })
        );
        //const videoStream = document.createElement("video");
        videoStream.muted = true;
        videoStream.onresize = e =>
          console.log(
            e.type,
            'every resize',
            e.target.videoWidth,
            e.target.videoHeight
          );
        videoStream.onloadedmetadata = e => {
          const { videoWidth, videoHeight } = e.target;
          console.log(e.type, videoWidth, videoHeight);
        };
        const ms = videoStream.captureStream();
        ms.onaddtrack = async e => {
          if (e.track.kind === 'audio') {
            await audioSender.replaceTrack(e.track);
          } else {
            await videoSender.replaceTrack(e.track);
            console.log(e.track.getSettings());
          }
        };
        for (const blobURL of urls) {
          await new Promise(resolve => {
            videoStream.addEventListener(
              'pause',
              async _ => {
                resolve();
              },
              {
                once: true,
              }
            );
            videoStream.addEventListener(
              'canplay',
              async _ => {
                if (!recorder) {
                  tracks = await tracks;
                  handleChromiumVideoElementMediaRecorderBug = document.createElement("video").srcObject = tracks;
                  recorder = new MediaRecorder(tracks, {
                    mimeType: 'video/webm',
                  });
                  recorder.onstart = _ => {
                    console.log(_.type);
                  };
                  result = new Promise(resolve => {
                    recorder.addEventListener(
                      'dataavailable',
                      ({ data }) => {
                        resolve(data);
                      },
                      {
                        once: true,
                      }
                    );
                  });
                  recorder.start();
                }
                await videoStream.play();
              },
              {
                once: true,
              }
            );
            videoStream.src = blobURL;
          });
        }
        recorder.stop();
        let blob = await result;
        let blobURL = URL.createObjectURL(blob);
        console.log(blob);
        await audioContext.close();
        worker.terminate();
        [mediaStream.getTracks(), stream.getTracks(), ms.getTracks()]
          .flat()
          .forEach(track => {
            track.enabled = false;
            track.stop();
            console.log(track);
          });
        video.srcObject = null;
        video.remove();
        videoStream.src = '';
        videoStream.remove();
        const recordedVideo = document.createElement('video');
        document.body.appendChild(recordedVideo);
        recordedVideo.autoplay = recordedVideo.controls = true;
        recordedVideo.onresize = ({ target: { videoWidth, videoHeight } }) => {
          console.log("resulting video resize", videoWidth, videoHeight);
        };
        recordedVideo.src = blobURL;
      })();
    </script>
  </body>
</html>
```

Reply:
Particularly when the input is a `MediaStream` from WebRTC the first `resize` events can be affected, which can be observed by attaching a `resize` event to `handleChromiumVideoElementMediaRecorderBug`

```
handleChromiumVideoElementMediaRecorderBug = document.createElement("video");
handleChromiumVideoElementMediaRecorderBug.srcObject = tracks;
handleChromiumVideoElementMediaRecorderBug.onresize = e =>
  console.log(
     e.type,
      'handleChromiumVideoElementMediaRecorderBug resize ' + new Date().getTime(),
       e.target.videoWidth,
       e.target.videoHeight
   );
/*
(2) resize handleChromiumVideoElementMediaRecorderBug resize 1573516555946 288 216
resize handleChromiumVideoElementMediaRecorderBug resize 1573516559885 384 288
..
resulting video resize 288 216
resulting video resize 384 288
*/
```

 Will adjust the code for Firefox, which does not currently fully support `AudioWorkletProcessor` to compare (see also https://github.com/w3c/mediacapture-record/issues/173).

Reply:
Even when the expected `videoWidth` and `videoHeight` are awaited at N `resize` events

```
                handleChromiumVideoElementMediaRecorderBug.onresize = async e => {
                  console.log(e.type, "another resize", handleChromiumVideoElementMediaRecorderBug.width, handleChromiumVideoElementMediaRecorderBug.height, handleChromiumVideoElementMediaRecorderBug.videoWidth,
                    handleChromiumVideoElementMediaRecorderBug.videoHeight);
                  if (handleChromiumVideoElementMediaRecorderBug.videoWidth === handleChromiumVideoElementMediaRecorderBug.srcObject.getVideoTracks()[0].width && handleChromiumVideoElementMediaRecorderBug.eight === getVideoTracks()[0].height) {
                  console.log(handleChromiumVideoElementMediaRecorderBug.videoWidth, handleChromiumVideoElementMediaRecorderBug.videoHeight);
                     handleChromiumVideoElementMediaRecorderBug.play();
                     videoStream.play()
                     if (recorder.state === "inactive") {
                       recorder.start();
                     }
                  }
                }
// see three resize events before expected values
// resize another resize 0 0 384 288
// resize another resize 0 0 576 432
// resize another resize 0 0 768 576 // expected pixel dimensions
```

when `autoplay` is not set other issues can be encountered, e.g., the video might not proceed with playback. 

Therefore, the test is not as straightforward in the field as might be expected relevant to the ccurrent standard.

Reply:
The code for Firefox, Nightly is different in several ways, though relevant to a variable resolution video produced by `MediaRecorder` being played, after several tests even when the initial video being played does output expected result at Nightly 72 using VP8 codec

`768 576`

the `videoWidth` of the first `resize` event of the first encoded track is incorrect

```
720 576 
480 240 
640 360  
400 300  
1280 720
```

the first `width` value should be `768`. 

The linked version of the WebRTC code provides a means to demonstrate the edge cases re what is expected and actual result of `resize` event. For completeness of the tests it depends on what values are is being expected when specific codecs or procedures are used to create the video whether a pass is true or not.

Reply:
FWIW from testing the most reliable means that have found to reliably record variable resolution input and output expected values for `videoWidth` and `videoHeight` of the underlying encoded video tracks at the _first_ `resize` event of HTML `<video>` element at Firefox, Nightly https://github.com/guest271314/MediaFragmentRecorder/blob/master/MediaFragmentRecorder.html for Chromium https://github.com/guest271314/MediaFragmentRecorder/blob/imagecapture-audiocontext-readablestream-writablestream/MediaFragmentRecorder.html.

Reply:
Let's go back to what the test in this PR is doing. It's loading a video from a file by setting `src`.

The spec fires `resize` in two places: in the resource fetch algorithm, and the "Whenever the intrinsic width or intrinsic height of the video changes". The latter doesn't apply during loading, since `readyState` will be `HAVE_NOTHING`.

So we know which part of the spec is responsible for firing the first `resize` event for this test. It's the part that I quoted earlier.

The remaining question is whether the media file's next resize is so soon that it happens before the first event `resize` event is handled. Is the resize the very next frame, or is it a few seconds in? If the latter, the test should be OK.

The HTML media API can get things our of sync. The `resize` event and videoWidth/videoHeight is but one aspect where this shines through.

Reply:
@zcorpan Posted files mentioned by @foolip for testing at https://github.com/web-platform-tests/wpt/pull/19030#issuecomment-551857047. Described edge cases at https://github.com/web-platform-tests/wpt/pull/19030#issuecomment-551861558 re expectations of the test and actual implementations. At https://github.com/web-platform-tests/wpt/pull/19030#discussion_r344339801 asked where the term "first" is actually used in the specification, to avoid either broad or narrow _interpretation_ of the current specification, where the language "first" is not present in the specification, right now. 

Edge cases exist which can output values that can make the test not pass, even where the eventual `videoWidth` and `videoHeight` could be set to the expected values at subsequent `resize` events, and that the language "first", as used in the current test, does not appear in the current referenced standard. 

Are you currently required to rely only on your interpretation of the language in the standard which you believe implies at the "first" `resize` event the `videoWidth` and `videoHeight` MUST reflect the underlying video frame; for example, where the current frame is captured by a separate application, at precisely T time during playback, and compared to the value at the current `resize` event?

Are you able to cite the specific language in the current HTML standard where the term "first" is used relevant to expected results of `videoWidth` and `videoHeight` values at a `resize` event?

Reading the test in isolation, then corss-referencing the standard, would search for the term "first" re the `resize` event, and such language cannot currently be found.

Is "first" necessary, where that language does not actually appear in the standard?

Consider the case of filing a Chromium or Firefox bug, and state that the `videoWidth` and `videoHeight` at the "first" `resize` event MUST reflect the pixel dimensions of the underlying frame, could not the reviewer of the issue reasonably ask: "Where are the terms 'first' and 'MUST' in the HTML Standard?". That is the concerned raised. Is the text in the current test 

`after first`

accurate and necessary? 

Given there are edge cases and the actual standard being referenced does not contain that language.

If, from your perspective, "first" and MUST are implied by the current standard, even though that language is not present, and the edge cases can be tested in other tests, or at least not omitted from consideration so developers are not expecting any edge cases exist re `resize`, you are free to utilize the videos in the test as you see fit, however, noting here that do not concur with such an interpretation of the current standard, where language cannot be ex post facto inserted into any primary source record.

Reply:
@fooip @zcorpan 

Am not certain about the reason the test is failing though a test _can_ fail when the initial `resize` event listener is not removed. Meaning `400`, `300` can be result though `200`, `150` is expected.

For the test to pass either previously attached `resize` event can be removed or `return` from the event handler

```
<!DOCTYPE html>
<html>

<head>
</head>

<body>
  <input type="file" multiple>
  <script>
    const input = document.querySelector("input[type=file]");
    input.onchange = async e => {
      let variableResolution = 0;
      for (const [index, file] of Object.entries(e.target.files)) {
        // await new Promise(resolve => {
          const video = document.createElement('video');
          // TODO: skip subtests after https://github.com/web-platform-tests/rfcs/pull/16
          if (video.canPlayType(file.type)) {
            // Load the video and wait for initial resize event.
            video.src = URL.createObjectURL(file);
            video.preload = 'auto';
            document.body.appendChild(video);
            video.onresize = e => {
              console.log(e, index);
              if (index == 0) {
                console.log(index, video.videoWidth === 400, video.videoHeight === 300);
              }
              if (index == 1 && variableResolution === 0) {
                if (variableResolution === 0) {
                  console.log(index, video.videoWidth, video.videoHeight);
                  console.assert(video.videoWidth === 400 && video.videoHeight === 300, [video.videoWidth, video.videoHeight]);
                  ++variableResolution;
                  // important if using same event handler
                  // test will not pass if `return` not present
                  return;
                }
              }
              if (index == 1 && variableResolution === 1) {
                console.log(index, video.videoWidth, video.videoHeight);
                console.assert(video.videoWidth === 200 && video.videoWidth, video.videoHeight === 150, [video.videoWidth, video.videoHeight]);
              }

            }

          }
          video.play();
        // video.onended = resolve;
        // });
      }
    }
  </script>
</body>

</html>
```

Reply:
It should be possible to modify the code at https://plnkr.co/edit/Dmsxt19ci9hZhscLjlR9?p=preview  for WPT tests. Tested at Chromium 80 and Firefox 69.

Reply:
@foolip 

> My interpretation is that the event order isn't reliably in the expected order, but I think this is a case where flaky fail is the best case.

Why is `playing` event included in the test?

Reply:
For an unknown reason the "next" version of plnkr does not proceed with playback following first two `resize` events. The "old" version does proceed with playback https://plnkr.co/edit/Dmsxt19ci9hZhscLjlR9?p=info, which outputs the expected order of events, where `resize` is fired first,  then `playing`, `resize`, `playing`, then `resize`.

Would rewrite the test, though am not certain how `EventWatcher` works.

Reply:
Got this result for the WPT test link for Safari (https://github.com/web-platform-tests/wpt/commit/03bcb6d910). If Safari 1) fetching the mp4 video 2) capable of playback of that particular media file?


> FAIL message: assert_equals: Expected resize event, but got ended event instead expected "resize" but got "ended"

Reply:
@foolip If the current test is only playing one video 

`video.src = `/media/400x300-red-resize-300x150-green.${format}`;`

why is `playing` event expected to be fired _twice_ at https://github.com/web-platform-tests/wpt/blob/03bcb6d91029419e7144850d72414aefa6d6b637/html/semantics/embedded-content/the-video-element/resize-during-playback.html#L24

`const eventWatcher = new EventWatcher(t, video, ['resize', 'playing', 'error', 'ended']);`

and https://github.com/web-platform-tests/wpt/blob/03bcb6d91029419e7144850d72414aefa6d6b637/html/semantics/embedded-content/the-video-element/resize-during-playback.html#L34

`await eventWatcher.wait_for(['playing', 'resize']);`

where the documentation for `EventWatcher` states

> `wait_for` either takes the name of a single event and returns a Promise that will resolve after that event is fired at the watched node, or else it takes an array of the names of a series of events and returns a Promise that will resolve **after that specific series of events has been fired at the watched node**. (emp. added)

and `play()` is executed once, `playing` should only expected to be fired once?

Reply:
If the MP4 file is fetched and playable at Safari (do not have access to Safari to test) the pattern below should output the expected events in the expected order (see version 2 https://plnkr.co/edit/Dmsxt19ci9hZhscLjlR9?p=preview)
```
<!DOCTYPE html>
<html>

<head>
  <title>Test video resize</title>
</head>

<body>
  <script>
    (async e => {
      const files = ["data:video/webm;base64,.."];

      for (const [index, file] of Object.entries(files)) {
        await new Promise(async resolve => {
            const blob = await (await fetch(file)).blob();
            const video = document.createElement('video');
            // TODO: skip subtests after https://github.com/web-platform-tests/rfcs/pull/16
            if (video.canPlayType(blob.type)) {
              // Load the video and wait for initial resize event.
              video.src = URL.createObjectURL(blob);
              video.preload = 'auto';
              document.body.appendChild(video);
              let changed = false;
              let first = false;
              const eventOrder = ["resize", "playing", "resize"];
              const order = [0, 1, 2];
              video.onended = _ => resolve(`Resize playback test: ${order.length === 0 ? "PASS" : "FAIL"}`);
              video.onresize = async e => {
                if (!changed && !first) {
                  if (e.type === eventOrder[order[0]] && (video.videoWidth === 400 && video.videoHeight === 300)) {
                    order.shift();
                  }
                  console.log(e.type, (video.videoWidth === 400 && video.videoHeight === 300), [video.videoWidth, video.videoHeight]);
                  changed = true;
                }
                if (changed && first) {
                  console.log(e.type, (video.videoWidth === 200 && video.videoHeight === 150), [video.videoWidth, video.videoHeight]);
                  if (e.type === eventOrder[order[0]] && (video.videoWidth === 200 && video.videoHeight === 150)) {
                    order.shift();
                  }
                }
                if (!first) {
                  first = true;
                  video.onplaying = e => {
                    console.log(e.type, (video.videoWidth === 400 && video.videoHeight === 300), [video.videoWidth, video.videoHeight]);
                    if (e.type === eventOrder[order[0]] && (video.videoWidth === 400 && video.videoHeight === 300)) {
                      order.shift();
                    }
                  }
                  try {
                    await video.play();
                  } catch (e) {
                    console.error(e);
                  }
                }
              }
            }
          })
          .then(result => console.log(result))
      }
    })();
  </script>
</body>

</html>
```

Reply:
The word "first" is not necessary. The assertion descriptions could be removed completely, but that makes debugging a failing test harder. "First" is there to make debugging easier -- of the two `resize` handlers in the test, one appears before the other.

> Are you currently required to rely only on your interpretation of the language in the standard which you believe implies at the "first" resize event the videoWidth and videoHeight MUST reflect the underlying video frame; for example, where the current frame is captured by a separate application, at precisely T time during playback, and compared to the value at the current resize event?
>
> Are you able to cite the specific language in the current HTML standard where the term "first" is used relevant to expected results of videoWidth and videoHeight values at a resize event?

I've already said that the value could be different if the video resizes again before the `resize` event is handled. If the test in this PR experiences that problem, the right fix is to delay the resize in the video so that the browser has enough time to fire the first `resize` event before the next resize happens.

It's not clear to me whether you think the test is wrong, or the spec is wrong, or if my interpretation of the spec is wrong? Or are you just concerned about the word "first"?

Reply:
@zcorpan 

> It's not clear to me whether you think the test is wrong, or the spec is wrong, or if my interpretation of the spec is wrong? Or are you just concerned about the word "first"?

The test handles the case of a `<video>` with `src` set to a media file. The test does not currently handle the case of a `<video>` with `srcObject` set to a `MediaStream`. Noting here to avoid the omission of that fact.

The spec could be improved relevant to the above described case re `srcObject` set to a `MediaStream`. Currently that case is not considered at the spec.

Would not use the term "wrong" re your interpretation of the current spec, though perhaps that is necessary to convey own interpretation of the spec: If the term is not present in the spec, and in fact any primary source document, inference of a specific action relevant to a term that is not present in the primary source document is not based on the record itself. If in fact the spec is mandating that to be in compliance with the spec implementers MUST set `videoWidth` and `videoHeight` and display the same precisely at the first `resize` event, that language needs to be included in the spec. 

Neither individuals nor institutions can arbitrarily infer specific meaning where the language inferred by that individual or institution is not actually present in the primary source document.

Reply:
Is there an issue filed about the srcObject case?

> If the term is not present in the spec, and in fact any primary source document, inference of a specific action relevant to a term that is not present in the primary source document is not based on the record itself.

I'm having difficulty understanding this.

> If in fact the spec is mandating that to be in compliance with the spec implementers MUST set videoWidth and videoHeight and display the same precisely at the first resize event, that language needs to be included in the spec.

It doesn't require the display to match up with the event timing, certainly.

I'm not sure this is leading anywhere. If you think the spec is missing something, the right place to explain that is in a new issue.

Reply:
> Is there an issue filed about the srcObject case?

No.

> I'm having difficulty understanding this.

The specification does not include the language "first" and "MUST" re first `resize` event. An individual inference that the current spec means that without the language actually being updating in the spec is lacks concrete citation for such an interpretation: Update the spec to say what is claimed as being inferred.

> I'm not sure this is leading anywhere. If you think the spec is missing something, the right place to explain that is in a new issue.

Am banned from contributing to WHATWG.